### ISSUE TITLE
Wrong Implementation of SimCLR

### ISSUE THREAD

**Issue by chingyaoc** [PyTorchLightning/lightning-bolts#695]
## 🐛 Bug

In line 253 of ```pl_bolts/models/self_supervised/simclr/simclr_module.py```, ```neg``` is subtract with e^(1/temp) to remove similarity measure for (x1,x1). However, this only guarantees that the scale of the loss is correct. There is still gradient backpropagating through the pair (x1,x1).

A correct implementation should be using a mask to eliminate the wrong negative pair. Here are some examples to implement the masking: [facebookresearch/vissl](https://github.com/facebookresearch/vissl/blob/master/vissl/losses/simclr_info_nce_loss.py#L145) and [chingyaoc/DCL](https://github.com/chingyaoc/DCL/blob/master/main.py#L37). After fixing this bug, I can improve the performance on CIFAR-10 from 88.5% (reported [here](https://lightning-bolts.readthedocs.io/en/latest/self_supervised_models.html)) to roughly 92%.

**Comment by ohayonguy** [comment]
Would this be correct then?

```
    def _get_negative_mask(self, sim, out):
        negative_mask = torch.ones_like(sim, dtype=torch.bool, device=self.device)
        for i in range(out.shape[0]):
            negative_mask[i, i] = False
        return negative_mask

    def nt_xent_loss(self, out_1, out_2, temperature, eps=1e-6):
        """
            assume out_1 and out_2 are normalized
            out_1: [batch_size, dim]
            out_2: [batch_size, dim]
        """
        # gather representations in case of distributed training
        # out_1_dist: [batch_size * world_size, dim]
        # out_2_dist: [batch_size * world_size, dim]
        if torch.distributed.is_available() and torch.distributed.is_initialized():
            out_1_dist = SyncFunction.apply(out_1)
            out_2_dist = SyncFunction.apply(out_2)
        else:
            out_1_dist = out_1
            out_2_dist = out_2

        # out: [2 * batch_size, dim]
        # out_dist: [2 * batch_size * world_size, dim]
        out = torch.cat([out_1, out_2], dim=0)
        out_dist = torch.cat([out_1_dist, out_2_dist], dim=0)

        # cov and sim: [2 * batch_size, 2 * batch_size * world_size]
        # neg: [2 * batch_size]
        cov = torch.mm(out, out_dist.t().contiguous())
        sim = torch.exp(cov / temperature)

        neg = (sim * self._get_negative_mask(sim, out)).sum(dim=-1)

        # Positive similarity, pos becomes [2 * batch_size]
        pos = torch.exp(torch.sum(out_1 * out_2, dim=-1) / temperature)
        pos = torch.cat([pos, pos], dim=0)

        loss = -torch.log(pos / (neg + eps)).mean()

        return loss
```
The mask can be generated beforehand to improve performance, of course.

@chingyaoc 
@Borda

**Comment by ananyahjha93** [comment]
We are deprecating the self-supervised learning paper implementations in bolts and instead provided a [VISSL](https://vissl.ai/) integration in PyTorch Lightning Flash which can be used to train models using self-supervised learning algorithms. You can refer to the documentation [here](https://lightning-flash.readthedocs.io/en/latest/reference/image_embedder.html).
